### PR TITLE
Change Dictionary.ContainsKey() + Dictionary[] combination to Dictionary.TryGetValue()

### PR DESCRIPTION
**Pull request by Arkatufus**
Change all Dictionary.ContainsKey(key) + Dictionary[key] combination to Dictionary.TryGetValue(key, out value)

### CODE REVIEW

**Comment by Silv3rcircl3** [comment]
@Arkatufus looks like some of the persistence tests failed, could you take a look at those ? http://petabridge-ci.cloudapp.net/viewLog.html?buildId=26036&buildTypeId=AkkaNet_AkkaNetWindowsUnitTests&guest=1

File: src/contrib/cluster/Akka.Cluster.Tools/Client/ClusterReceptionist.cs [hunk 1]
```diff
@@ -546,14 +546,14 @@ protected override bool Receive(object message)
 
         private void UpdateClientInteractions(IActorRef client)
         {
-            if (_clientInteractions.ContainsKey(client))
+            DeadlineFailureDetector failureDetector;
+            if (_clientInteractions.TryGetValue(client, out failureDetector))
```
Comment (Silv3rcircl3) [inline src/contrib/cluster/Akka.Cluster.Tools/Client/ClusterReceptionist.cs]: Curly braces can be removed

File: src/core/Akka.Cluster/VectorClock.cs [hunk 2]
```diff
@@ -439,9 +439,10 @@ public VectorClock Merge(VectorClock that)
         /// <returns>A newly created <see cref="VectorClock"/> that has the given node removed.</returns>
         public VectorClock Prune(Node removedNode)
         {
-            if (Versions.ContainsKey(removedNode))
+            var newVersions = Versions.Remove(removedNode);
+            if (!ReferenceEquals(newVersions, Versions))
```
Comment (Silv3rcircl3) [inline src/core/Akka.Cluster/VectorClock.cs]: Curly braces can be removed

File: src/core/Akka.Persistence/Fsm/PersistentFSMBase.cs [hunk 3]
```diff
@@ -282,15 +282,15 @@ public void SetTimer(string name, object msg, TimeSpan timeout, bool repeat = fa
         {
             if (DebugEvent)
                 _log.Debug("setting " + (repeat ? "repeating" : "") + "timer '{0}' / {1}: {2}", name, timeout, msg);
-            if (_timers.ContainsKey(name))
-                _timers[name].Cancel();
+
+            Timer oldTimer;
+            if (_timers.TryGetValue(name, out oldTimer)) {
```
Comment (Silv3rcircl3) [inline src/core/Akka.Persistence/Fsm/PersistentFSMBase.cs]: Curly braces can be removed

File: src/core/Akka.Remote/DefaultFailureDetectorRegistry.cs [hunk 4]
```diff
@@ -51,7 +51,12 @@ public DefaultFailureDetectorRegistry(Func<FailureDetector> factory)
         /// <returns>TBD</returns>
         public bool IsAvailable(T resource)
         {
-            return !ResourceToFailureDetector.ContainsKey(resource) || ResourceToFailureDetector[resource].IsAvailable;
+            FailureDetector failureDetector;
+            if (ResourceToFailureDetector.TryGetValue(resource, out failureDetector))
```
Comment (Silv3rcircl3) [inline src/core/Akka.Remote/DefaultFailureDetectorRegistry.cs]: Curly braces can be removed

File: src/core/Akka.Remote/DefaultFailureDetectorRegistry.cs [hunk 5]
```diff
@@ -61,7 +66,12 @@ public bool IsAvailable(T resource)
         /// <returns>TBD</returns>
         public bool IsMonitoring(T resource)
         {
-            return ResourceToFailureDetector.ContainsKey(resource) && ResourceToFailureDetector[resource].IsMonitoring;
+            FailureDetector failureDetector;
+            if (ResourceToFailureDetector.TryGetValue(resource, out failureDetector))
+            {
```
Comment (Silv3rcircl3) [inline src/core/Akka.Remote/DefaultFailureDetectorRegistry.cs]: Curly braces can be removed

File: src/core/Akka.Remote/DefaultFailureDetectorRegistry.cs [hunk 6]
```diff
@@ -70,9 +80,10 @@ public bool IsMonitoring(T resource)
         /// <param name="resource">TBD</param>
         public void Heartbeat(T resource)
         {
-            if (ResourceToFailureDetector.ContainsKey(resource))
+            FailureDetector failureDetector;
+            if (ResourceToFailureDetector.TryGetValue(resource, out failureDetector))
```
Comment (Silv3rcircl3) [inline src/core/Akka.Remote/DefaultFailureDetectorRegistry.cs]: Curly braces can be removed

File: src/core/Akka.Remote/DefaultFailureDetectorRegistry.cs [hunk 7]
```diff
@@ -82,9 +93,9 @@ public void Heartbeat(T resource)
                     // First check for non-existing key wa outside the lock, and a second thread might just have released the lock
                     // when this one acquired it, so the second check is needed (double-check locking pattern)
                     var oldTable = new Dictionary<T, FailureDetector>(ResourceToFailureDetector);
-                    if (oldTable.ContainsKey(resource))
+                    if (oldTable.TryGetValue(resource, out failureDetector))
```
Comment (Silv3rcircl3) [inline src/core/Akka.Remote/DefaultFailureDetectorRegistry.cs]: Curly braces can be removed

File: src/core/Akka.Remote/EndpointManager.cs [hunk 8]
```diff
@@ -1016,7 +1016,9 @@ private void HandleInboundAssociation(InboundAssociation ia, bool writerIsIdle)
             if (readonlyEndpoint != null)
             {
                 var endpoint = readonlyEndpoint.Item1;
-                if (_pendingReadHandoffs.ContainsKey(endpoint)) _pendingReadHandoffs[endpoint].Disassociate();
+                AkkaProtocolHandle protocolHandle;
+                if (_pendingReadHandoffs.TryGetValue(endpoint, out protocolHandle)) protocolHandle.Disassociate();
```
Comment (Silv3rcircl3) [inline src/core/Akka.Remote/EndpointManager.cs]: `protocolHandle.Disassociate();`underneath the `if`

File: src/core/Akka.Remote/EndpointManager.cs [hunk 9]
```diff
@@ -1169,8 +1171,9 @@ private void AcceptPendingReader(IActorRef takingOverFrom)
 
         private void RemovePendingReader(IActorRef takingOverFrom, AkkaProtocolHandle withHandle)
         {
-            if (_pendingReadHandoffs.ContainsKey(takingOverFrom) &&
-                _pendingReadHandoffs[takingOverFrom].Equals(withHandle))
+            AkkaProtocolHandle handle;
+            if (_pendingReadHandoffs.TryGetValue(takingOverFrom, out handle) &&
+                handle.Equals(withHandle))
```
Comment (Silv3rcircl3) [inline src/core/Akka.Remote/EndpointManager.cs]: Curly braces can be removed

File: src/core/Akka.Remote/RemoteWatcher.cs [hunk 10]
```diff
@@ -489,8 +489,17 @@ private void ReceiveHeartbeatRsp(int uid)
 
             if (WatcheeByNodes.ContainsKey(from) && !Unreachable.Contains(from))
             {
-                if (!_addressUids.ContainsKey(from) || _addressUids[from] != uid)
+                int addressUid;
+                if (_addressUids.TryGetValue(from, out addressUid))
```
Comment (Silv3rcircl3) [inline src/core/Akka.Remote/RemoteWatcher.cs]: I would like to keep it in one simple `if` statement, i.e.

```csharp
if (!_addressUids.TryGetValue(from, out int addressUid) || adressUid != uid)
```

File: src/core/Akka/Configuration/Hocon/HoconObject.cs [hunk 11]
```diff
@@ -188,12 +190,11 @@ public void Merge(HoconObject other)
 
             foreach (var otherItem in otherItems)
             {
-                if (thisItems.ContainsKey(otherItem.Key))
+                //if other key was present in this object and if we have a value, 
+                //just ignore the other value, unless it is an object
```
Comment (Silv3rcircl3) [inline src/core/Akka/Configuration/Hocon/HoconObject.cs]: :+1:

**Comment by Silv3rcircl3** [comment]
Nice work @Arkatufus, have only some code formatting remarks 👍

File: src/core/Akka.Persistence/Fsm/PersistentFSMBase.cs [hunk 12]
```diff
@@ -332,10 +333,7 @@ public bool IsTimerActive(string name)
         /// <param name="timeout">TBD</param>
         public void SetStateTimeout(TState state, TimeSpan? timeout)
         {
-            if (!_stateTimeouts.ContainsKey(state))
-                _stateTimeouts.Add(state, timeout);
-            else
-                _stateTimeouts[state] = timeout;
+            _stateTimeouts[state] = timeout;
```
Comment (Arkatufus) [inline src/core/Akka.Persistence/Fsm/PersistentFSMBase.cs]: Do you want to use expression body everywhere or just for this instance?

File: src/contrib/cluster/Akka.Cluster.Sharding/PersistentShardCoordinator.cs [hunk 13]
```diff
@@ -468,7 +468,8 @@ private void HandleRebalanceDone(RebalanceDone done)
 
             IActorRef region;
             // The shard could have been removed by ShardRegionTerminated
-            if (_currentState.Shards.ContainsKey(done.Shard))
+
+            if (_currentState.Shards.TryGetValue(done.Shard, out region))
```
Comment (Arkatufus) [inline src/contrib/cluster/Akka.Cluster.Sharding/PersistentShardCoordinator.cs]: I think this style have a "scoping problem" and can make a code a bit hard to read?
I can probably use it if the variable in question isn't used outside the scope?

Comment (Silv3rcircl3) [inline src/contrib/cluster/Akka.Cluster.Sharding/PersistentShardCoordinator.cs]: Yup, quoting myself, "where possible" ;-)

File: src/core/Akka.Persistence/Fsm/PersistentFSMBase.cs (hunk 12 above)
Comment (Silv3rcircl3) [inline src/core/Akka.Persistence/Fsm/PersistentFSMBase.cs]: That was the only method I have seen that was affected from your changes, you don't need to do it for all places that aren't affected by your code

File: src/core/Akka.Remote/RemoteWatcher.cs (hunk 10 above)
Comment (Arkatufus) [inline src/core/Akka.Remote/RemoteWatcher.cs]: That could cause a bug in certain circumstances, I think. 
I'm not sure, but I'd like to keep it like that for now.

File: src/contrib/cluster/Akka.Cluster.Sharding/PersistentShard.cs [hunk 14]
```diff
@@ -135,12 +135,10 @@ protected void SaveSnapshotIfNeeded()
...
         protected override void EntityTerminated(IActorRef tref)
         {
-            ShardId id;
-            IImmutableList<Tuple<Msg, IActorRef>> buffer;
 
-            if (IdByRef.TryGetValue(tref, out id))
+            if (IdByRef.TryGetValue(tref, out ShardId id))
             {
-                if (MessageBuffers.TryGetValue(id, out buffer) && buffer.Count != 0)
+                if (MessageBuffers.TryGetValue(id, out IImmutableList<Tuple<Msg, IActorRef>> buffer) && buffer.Count != 0)
```
Comment (Silv3rcircl3) [inline src/contrib/cluster/Akka.Cluster.Sharding/PersistentShard.cs]: That escalated quickly 😄 
Could you replace those types with var ?

File: src/core/Akka.Remote/RemoteWatcher.cs [hunk 15]
```diff
@@ -479,18 +479,20 @@ private void ReceiveHeartbeatRsp(int uid)
...
-            {
                 Log.Debug("Received first heartbeat rsp from [{0}]", from);
-            }
 
             if (WatcheeByNodes.ContainsKey(from) && !Unreachable.Contains(from))
             {
-                if (!_addressUids.ContainsKey(from) || _addressUids[from] != uid)
+                if (_addressUids.TryGetValue(from, out int addressUid))
+                {
+                    if (addressUid == uid)
```
Comment (Silv3rcircl3) [inline src/core/Akka.Remote/RemoteWatcher.cs]: I think that needs to be != instead of ==

**Comment by Silv3rcircl3** [comment]
This is ready to go once the last remarks are fixed 👍

**Comment by Arkatufus** [comment]
Changes pushed to fork, hope everything's OK

**Comment by Aaronontheweb** [comment]
Re-running the specs since @heynickc has gotten the build system to run these stages properly since they were last tried.

**Comment by Aaronontheweb** [comment]
cc @heynickc appears to be an issue with Mono build - can you take a look?

**Comment by heynickc** [comment]
@Aaronontheweb looks like this branch needs to be rebased to our latest changes in v1.3.  Missing this line in build.sh https://github.com/akkadotnet/akka.net/blob/v1.3/build.sh#L89

**Comment by Silv3rcircl3** [comment]
@Arkatufus can you debase please?

**Comment by Aaronontheweb** [comment]
Took care of it

**Pull request merged**